rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala

```diff
@@ -0,0 +1,239 @@
+    Receive(
+      for { (xs, chan) <- term.binds } yield {
+        substitute(chan, env) match {
+          case Left(_chan) => (xs, _chan)
+          case Right(par)  => throw new Error(s"Cannot receive on expression: $par")
+        }
+      },
+      substitute(term.body, env),
+      term.persistent,
+      term.count
+    )
+
+  def rename(term: Receive, j: Int): Receive =
+    Receive(
+      for ((xs, channel) <- term.binds) yield {
+        (xs, rename(channel, j))
+      },
+      rename(term.body, j),
+      term.persistent,
+      term.count
+    )
+
+  def substitute(term: Var, env: Env): Either[Var, Data] =
+    term match {
+      case BoundVar(level) =>
+        env.get(level) match {
+          case Some(Left(quote)) => Right(Left(rename(quote, env.level)))
+          case Some(Right(par))  => Right(Right(rename(par, env.level)))
+          case None =>
+            if (level - env.level < 0) Left(BoundVar(0))
+            else Left(BoundVar(level - env.level))
+        }
+      case _ => Left(term)
+    }
+
+  def rename(term: Var, j: Int): Var =
+    term match {
+      case BoundVar(i) =>
+        if (i + j < 0) BoundVar(0)
+        else BoundVar(i + j)
+      case _ => term
+    }
+
+  def rename(term: Expr, j: Int): Expr =
+    term match {
+      case EVar(v: Var)             => EVar(rename(v, j))
+      case ENot(p: Par)             => ENot(rename(p, j))
+      case ENeg(p: Par)             => ENeg(rename(p, j))
+      case EMult(p1: Par, p2: Par)  => EMult(rename(p1, j), rename(p2, j))
+      case EDiv(p1: Par, p2: Par)   => EDiv(rename(p1, j), rename(p2, j))
+      case EPlus(p1: Par, p2: Par)  => EPlus(rename(p1, j), rename(p2, j))
+      case EMinus(p1: Par, p2: Par) => EMinus(rename(p1, j), rename(p2, j))
+      case ELt(p1: Par, p2: Par)    => ELt(rename(p1, j), rename(p2, j))
+      case ELte(p1: Par, p2: Par)   => ELte(rename(p1, j), rename(p2, j))
+      case EGt(p1: Par, p2: Par)    => EGt(rename(p1, j), rename(p2, j))
+      case EGte(p1: Par, p2: Par)   => EGte(rename(p1, j), rename(p2, j))
+      case EEq(p1: Par, p2: Par)    => EEq(rename(p1, j), rename(p2, j))
+      case ENeq(p1: Par, p2: Par)   => ENeq(rename(p1, j), rename(p2, j))
+      case EAnd(p1: Par, p2: Par)   => EAnd(rename(p1, j), rename(p2, j))
+      case EOr(p1: Par, p2: Par)    => EOr(rename(p1, j), rename(p2, j))
+    }
+
+  def rename(exp: Ground, j: Int): Ground =
+    exp match {
+      case EList(xs) =>
+        EList(for { par <- xs } yield {
+          rename(par, j)
+        })
+      case ETuple(xs) =>
+        ETuple(for { par <- xs } yield {
+          rename(par, j)
+        })
+      case ESet(xs) =>
+        ESet(for { par <- xs } yield {
+          rename(par, j)
+        })
+      case EMap(xs) =>
+        EMap(for { (par0, par1) <- xs } yield {
+          (rename(par0, j), rename(par1, j))
+        })
+      case g @ _ => g
+    }
+
+  def substitute(exp: EVar, env: Env): Either[Var, Data] =
```

Comment:
The return type here suggests that you should explicitly handle EVar differently.

Reply:
This is what my commit message was about. I don't see a way around the fact that a variable may be substituted either for a channel, or for a process, or not at all if there is no binding. I believe the return type is incorrect. It should be `Either[EVar,Data]`, but that makes little difference.

Reply:
What it's substituted for depends solely on where it occurs.
I don't think this should be called "substitute" it breaks symmetry with all the other substitute methods.